### PR TITLE
✨ Allow TLS minimum version to be configured

### LINKED ISSUE #1431: Default TLS implementations leave consuming systems open to multiple vulnerabilities.
The default implementation for TLS endpoints expose the long deprecated TLS 1.0 and TLS 1.1 versions along with dangerous ciphers that expose systems to version downgrade attacks, SWEET32 (CVE-2016-2183, CVE-2016-6329), BEAST (CVE-2011-3389), LUCKY13 (CVE-2013-0169).  There should be a way for consumers to control which versions and ciphers are exposed.

This was confirmed with a default endpoint and tested with version 3.0.4 of testssl.sh.

### PR DESCRIPTION
**Pull request by willthames**
Some environments have automated security scans that trigger
on TLS versions or insecure cipher suites. Setting TLS to 1.3
would solve both problems (setting to 1.2 only solves the former
as the default 1.2 cipher suites are insecure).

Default TLS minimum version of 1.0 remains.

### CODE REVIEW

**Comment by willthames** [comment]
One could argue that only 1.3 makes any real sense but I've gone with the same TLS version options as go's crypto package

**Comment by coderanger** [comment]
/ok-to-test

@willthames You'll need to rebase/amend to remove the `Fixes` from the commit message. Because of how Kubernetes vendoring works, it can get noisy to leave those in the commits themselves.

As for default version, I would be fine setting it to 1.3 if all recent kube-apiservers support that. That's the only thing which actually matters as a client.

**Comment by willthames** [comment]
See #1431 for related issue

**Comment by willthames** [comment]
@coderanger TLS 1.3 support was enabled by default with go 1.13 (https://golang.org/doc/go1.13#tls_1_3) but was available in go 1.12 through a GODEBUG setting. 

Looks like kubernetes builds started to use go 1.13 from kubernetes 1.17 (from https://github.com/kubernetes/kubernetes/commit/e3ff39ffa068bbd1d9808c54923e527248e4fa87)

AWS still supports 1.16 but kubernetes itself doesn't.

If you're happy for me to make 1.3 the default I can make that change (it'll make this PR a lot simpler!)

**Comment by coderanger** [comment]
I would want to make sure that's okay with OperatorSDK crew because they have to support older stuff than the rest of us (usually) but I think 1.2 as the default would definitely make sense no matter what, and maybe 1.3 if we can swing it. I think the overall structure of your change is good though, just playing with which is the default value.

File: pkg/webhook/server.go [hunk 1]
```diff
@@ -109,6 +117,17 @@ func (s *Server) setDefaults() {
...
 	}
+
+	switch s.TLSMinVersion {
+	case "1.1":
+		s.tlsMinVersion = tls.VersionTLS11
+	case "1.2":
+		s.tlsMinVersion = tls.VersionTLS12
+	case "1.3":
+		s.tlsMinVersion = tls.VersionTLS13
+	default:
```
Comment (alvaroaleman) [inline pkg/webhook/server.go:128]: IMHO we should error if `TLSMinVersion` is set but not in the list, silently picking an old version is not a good idea IMHO

**Comment by alvaroaleman** [comment]
Also nit, but I don't think this qualifies as a bugfix, a bug is IMHO something that was already built and supposed to work but didn't, since we never head support for configuring the TLS version this is IMHO a feature.

**Comment by willthames** [comment]
@alvaroaleman the description for bug also says patch (although I guess any change is a patch), and there is an issue raised for this - that was my reasoning but I'm not against changing it to feature

File: pkg/webhook/server.go (hunk 1 above)
Comment (willthames) [reply pkg/webhook/server.go:128]: Do you mind suggesting a mechanism for this (setDefaults doesn't currently have any error handling so I'm not sure what pattern to follow)

**Comment by willthames** [comment]
@alvaroaleman changed title to use ✨

File: pkg/webhook/server.go (hunk 1 above)
Comment (alvaroaleman) [reply pkg/webhook/server.go:128]: I would just move this into `Start`. The only thing that really qualifies as defaulting is to set it to `tls.VersionTLS10` if s.TLSMinVersion is an empty string

Comment (willthames) [reply pkg/webhook/server.go]: thanks for that suggestion - hopefully this iteration does what you're after.

**Review by alvaroaleman — APPROVED** [review]

**Comment by alvaroaleman** [comment]
thanks!

**Comment by coderanger** [comment]
@alvaroaleman Is there a reason we didn't make 1.2 the default?

**Comment by alvaroaleman** [comment]
> @alvaroaleman Is there a reason we didn't make 1.2 the default?

Yeah, to not change what is currently happening, golang defaults to 1.0:

```
 // If zero, TLS 1.0 is currently taken as the minimum.
```

**Comment by coderanger** [comment]
Sure but that seems like a bad default because kube-apiserver should support 1.2 in all cases? 1.3 was questionable but I don't think there's a reason to not pin up to 1.2 to block some funky MITM attacks?

**Comment by alvaroaleman** [comment]
> Sure but that seems like a bad default because kube-apiserver should support 1.2 in all cases? 1.3 was questionable but I don't think there's a reason to not pin up to 1.2 to block some funky MITM attacks?

So I am not opposed to changing the default, golang defaulting to 1.0 is quite terrible actually. It just happens to be a change that _might_ cause issues (I don't know if it does) and this change right here seemed to me just about adding a knob.

**Comment by coderanger** [comment]
Legit, I'll open a new PR to debate changing the default :)

**Pull request merged**